### PR TITLE
Random Refactoring

### PR DESCRIPTION
**Pull request by geoffthemedio**
-Replaces `std::random_shuffle`, which is removed in C++17
-Moves all access of the random generator behind mutex locks
-Replaces `boost` random generator and distributions with `std` equivalents

### CODE REVIEW

File: util/Random.cpp [hunk 1]
```diff
@@ -1,56 +1,62 @@
...
+    gen.seed(static_cast<GeneratorType::result_type>(diff.total_milliseconds()));
 }
 
-SmallIntDistType SmallIntDist(int min, int max) {
-    boost::mutex::scoped_lock lock(s_prng_mutex);
-    return SmallIntDistType(gen, boost::uniform_smallint<>(min, max));
+int RandInt(int min, int max) {
+    if (min <= max)
+        return min;
+    std::uniform_int_distribution<> dis(min, max);
```
Comment (adrianbroher) [inline util/Random.cpp]: C++11 does not guarrantee reproducible distributions accross implementations:

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2059r0.pdf

File: util/Random.cpp [hunk 2]
```diff
@@ -1,56 +1,62 @@
...
-IntDistType IntDist(int min, int max) {
-    boost::mutex::scoped_lock lock(s_prng_mutex);
-    return IntDistType(gen, boost::uniform_int<>(min, max));
+double RandZeroToOne()
+{ return RandDouble(0.0, 1.0); }
+
+double RandDouble(double min, double max) {
+    if (min <= max)
+        return min;
+    std::uniform_real_distribution<> dis(min, max);
```
Comment (adrianbroher) [inline util/Random.cpp]: Same applies here

File: util/Random.cpp [hunk 3]
```diff
@@ -1,56 +1,62 @@
...
+    }
 }
 
-DoubleDistType DoubleDist(double min, double max) {
-    boost::mutex::scoped_lock lock(s_prng_mutex);
-    return DoubleDistType(gen, boost::uniform_real<>(min, max));
+double RandGaussian(double mean, double sigma) {
+    if (sigma <= 0.0)
+        return mean;
+    std::normal_distribution<> dis(mean, sigma);
```
Comment (adrianbroher) [inline util/Random.cpp]: And here

**Comment by adrianbroher** [comment]
Seems like you can combine the boost random distribution with the std generator:

https://github.com/Orphis/RandoBlazer/blob/master/Random.cpp

**Comment by Cjkjvfnby** [comment]
Could you update docstring for Random module with information about the distribution and stability of the result on different platforms?

>  -Moves all access of the random generator behind mutex locks

How may random calls do we have? Does it affect performance?

**Comment by geoffthemedio** [comment]
> > -Moves all access of the random generator behind mutex locks
> 
> How may random calls do we have?

I found 26 `RandInt` calls, 12 `RandZeroToOne` calls, and 7 `RandDouble` calls in C++ code. Some of those are in loops over objects in a set, and how often some are used depends on scripted content and Python..

> Does it affect performance?

Probably not much. The previous implementation was using similar locks in most of the same places, but in a way that didn't actually always usefully guard the generator in some cases.

Some during universe generation, some (probably unnecessary) during some `Condition` evaluations, and a few in explici

**Comment by geoffthemedio** [comment]
reworked using the boost distributions

File: UI/ClientUI.h [hunk 4]
```diff
@@ -234,7 +226,8 @@ class ClientUI {
     std::shared_ptr<SaveFileDialog>         m_savefile_dialog;
     std::shared_ptr<PasswordEnterWnd>       m_password_enter_wnd;   //!< the authentication window
 
-    PrefixedTextures                        m_prefixed_textures;
+    std::map<std::string, std::vector<std::shared_ptr<GG::Texture>>>
```
Comment (adrianbroher) [inline UI/ClientUI.h:233]: Would be useful to document what the key refers to. Filename? Path? Prefix? Something else?

**Comment by adrianbroher** [comment]
LGTM otherwise.

**Pull request merged**